modules/common/math/mpc_osqp.cc

```diff
@@ -0,0 +1,277 @@
+    for (int j = 0; j < state_dim_; ++j) {
+      // (row, val)
+      columns[i * state_dim_ + j].emplace_back(i * state_dim_ + j,
+                                               matrix_q_(j, j));
+      ++value_index;
+    }
+  }
+  // control
+  const int state_total_dim = state_dim_ * (horizon_ + 1);
+  for (int i = 0; i < horizon_; ++i) {
+    for (int j = 0; j < control_dim_; ++j) {
+      // (row, val)
+      columns[i * control_dim_ + j + state_total_dim].emplace_back(
+          state_total_dim + i * control_dim_ + j, matrix_r_(j, j));
+      ++value_index;
+    }
+  }
+  CHECK_EQ(value_index, num_param_);
+
+  int ind_p = 0;
+  for (int i = 0; i < num_param_; ++i) {
+    // TODO(SHU) Check this
+    P_indptr->push_back(ind_p);
+    for (const auto &row_data_pair : columns[i]) {
+      P_data->push_back(row_data_pair.second);    // val
+      P_indices->push_back(row_data_pair.first);  // row
+      ++ind_p;
+    }
+  }
+  P_indptr->push_back(ind_p);
+}
+
+// reference is always zero
+void MpcOsqp::castMPCToQPGradient() {
+  // populate the gradient vector
+  gradient_ = Eigen::VectorXd::Zero(
+      state_dim_ * (horizon_ + 1) + control_dim_ * horizon_, 1);
+}
+
+// equality constraints x(k+1) = A*x(k)
+void MpcOsqp::CalculateEqualityConstraint(std::vector<c_float> *A_data,
+                                          std::vector<c_int> *A_indices,
+                                          std::vector<c_int> *A_indptr) {
+  constexpr double kEpsilon = 1e-6;
+  // block matrix
+  Eigen::MatrixXd matrix_constraint(
+      state_dim_ * (horizon_ + 1) + state_dim_ * (horizon_ + 1) +
+          control_dim_ * horizon_,
+      state_dim_ * (horizon_ + 1) + control_dim_ * horizon_);
+  Eigen::MatrixXd state_identity_mat = Eigen::MatrixXd::Identity(
+      state_dim_ * (horizon_ + 1), state_dim_ * (horizon_ + 1));
+
+  matrix_constraint.block(0, 0, state_dim_ * (horizon_ + 1),
+                          state_dim_ * (horizon_ + 1)) =
+      -1 * state_identity_mat;
+
+  Eigen::MatrixXd control_identity_mat(control_dim_, control_dim_);
+
+  for (int i = 0; i < horizon_; i++) {
+    matrix_constraint.block((i + 1) * state_dim_, i * state_dim_, state_dim_,
+                            state_dim_) = matrix_a_;
+  }
+
+  for (int i = 0; i < horizon_; i++) {
+    matrix_constraint.block((i + 1) * state_dim_,
+                            i * control_dim_ + (horizon_ + 1) * state_dim_,
+                            state_dim_, control_dim_) = matrix_b_;
+  }
+  Eigen::MatrixXd all_identity_mat =
+      Eigen::MatrixXd::Identity(num_param_, num_param_);
+
+  matrix_constraint.block(state_dim_ * (horizon_ + 1), 0, num_param_,
+                          num_param_) = all_identity_mat;
+
+  std::vector<std::vector<std::pair<c_int, c_float>>> columns;
+  columns.resize(num_param_ + 1);  // TODO(SHU): double check
+  int value_index = 0;
+  // state and terminal state
+  for (int i = 0; i < num_param_; ++i) {                                // col
```

Comment:
Use size_t when applicable to avoid out of boundary violation for vectors.

Reply:
Fixed. Thank you.